src/background-filter.cpp

```diff
@@ -582,30 +542,47 @@ static void filter_video_render(void *data, gs_effect_t *_effect)
   }
   gs_stagesurface_unmap(tf->stagesurface);
 
-  if (tf->outputBGRA.empty() || static_cast<uint32_t>(tf->outputBGRA.cols) != width ||
-      static_cast<uint32_t>(tf->outputBGRA.rows) != height) {
+  // Output the masked image
+
+  if (!tf->effect) {
+    // Effect failed to load, skip rendering
     obs_source_skip_video_filter(tf->source);
     return;
   }
-  cv::Mat imageBGRA;
+
+  if (!obs_source_process_filter_begin(tf->source, GS_RGBA, OBS_ALLOW_DIRECT_RENDERING)) {
+    obs_source_skip_video_filter(tf->source);
+    return;
+  }
+
+  gs_texture_t *alphaTexture = nullptr;
   {
     std::lock_guard<std::mutex> lock(tf->outputBGRALock);
-    imageBGRA = tf->outputBGRA.clone();
+    alphaTexture = gs_texture_create(tf->backgroundMask.cols, tf->backgroundMask.rows, GS_R8, 1,
+                                     (const uint8_t **)&tf->backgroundMask.data, 0);
+    if (!alphaTexture) {
+      blog(LOG_ERROR, "Failed to create alpha texture");
+      obs_source_skip_video_filter(tf->source);
+      return;
+    }
   }
-  const uint8_t *textureData[] = {imageBGRA.data};
-  gs_texture_t *texture = gs_texture_create(width, height, GS_BGRA, 1, textureData, 0);
-  gs_effect_t *effect = obs_get_base_effect(OBS_EFFECT_DEFAULT);
-  gs_eparam_t *image = gs_effect_get_param_by_name(effect, "image");
-  gs_effect_set_texture(image, texture);
+  gs_eparam_t *param = gs_effect_get_param_by_name(tf->effect, "alphamask");
+  gs_effect_set_texture(param, alphaTexture);
+  param = gs_effect_get_param_by_name(tf->effect, "blurSize");
+  gs_effect_set_int(param, (int)tf->blurBackground);
+  param = gs_effect_get_param_by_name(tf->effect, "xTexelSize");
+  gs_effect_set_float(param, 1.0f / width);
+  param = gs_effect_get_param_by_name(tf->effect, "yTexelSize");
+  gs_effect_set_float(param, 1.0f / height);
+
```

Comment:
```suggestion
  gs_eparam_t *alphamask = gs_effect_get_param_by_name(tf->effect, "alphamask");
  gs_eparam_t *blurSize = gs_effect_get_param_by_name(tf->effect, "blurSize");
  gs_eparam_t *xTexelSize = gs_effect_get_param_by_name(tf->effect, "xTexelSize");
  gs_eparam_t *yTexelSize = gs_effect_get_param_by_name(tf->effect, "yTexelSize");

  gs_effect_set_texture(alphamask, alphamask_param);
  gs_effect_set_int(blurSize, (int)tf->blurBackground);
  gs_effect_set_float(xTexelSize, 1.0f / width);
  gs_effect_set_float(yTexelSize, 1.0f / height);

```

I think params can be stored on struct but I'm not sure